base-containers/base/inginious_container_api/utils.py

```diff
@@ -96,3 +105,213 @@ def ssh_wait(ssh_user, timeout=None):
+        buf += await reader.read(4 - len(buf))
+    length = struct.unpack('!I', bytes(buf))[0]
+    buf = bytearray()
+    while len(buf) != length and not reader.at_eof():
+        buf += await reader.read(length - len(buf))
+    message = msgpack.unpackb(bytes(buf), use_list=False)  # EXCEPTION EXTRA DATA
+    return message
+
+
+async def stdio():
+    """ Create the stdin and stdout streams to communicate with the agent """
+    my_loop = asyncio.get_event_loop()
+    reader = asyncio.StreamReader()
+    reader_protocol = asyncio.StreamReaderProtocol(reader)
+    writer_transport, writer_protocol = await my_loop.connect_write_pipe(asyncio.streams.FlowControlMixin, os.fdopen(1, 'wb'))
+    writer = asyncio.StreamWriter(writer_transport, writer_protocol, None, my_loop)
+    await my_loop.connect_read_pipe(lambda: reader_protocol, sys.stdin)
+    return reader, writer
+
+
+def handle_stdin_message(msg, proc_input, proc):
+    """ Process a single message from the agent (stdin message for the student code process or signals messages).
+    Used only when both containers are not on a shared kernel """
+    try:
+        if msg["type"] == "stdin":
+            input_content = msg["message"] + '\n'
+            proc_input.write(input_content.encode("utf8"))
+            proc_input.flush()
+            return "stdin ok"
+        if msg["type"] == "student_signal":
+            signal = msg["signal_data"]
+            proc.send_signal(int(signal.decode('utf8')))
+            return "signal ok"
+    except IOError as ioerror:
+        if ioerror.errno == errno.EPIPE:
+            return "pipe_closed"
+    except:
+        return
+
+
+async def handle_stdin(reader: asyncio.StreamReader, proc_input, proc):
+    """ Deamon to handle messages from the agent.
+    Used only when both containers are not on a shared kernel"""
+    try:
+        while not reader.at_eof():
+            buf = bytearray()
+            while len(buf) != 4 and not reader.at_eof():
+                buf += await reader.read(4 - len(buf))
+            if reader.at_eof():
+                continue
+            length = struct.unpack('!I', bytes(buf))[0]
+            buf = bytearray()
+            while len(buf) != length and not reader.at_eof():
+                buf += await reader.read(length - len(buf))
+            if reader.at_eof():
+                continue
+            message = msgpack.unpackb(bytes(buf), use_list=False)  # EXCEPTION EXTRA DATA
+            status = handle_stdin_message(message, proc_input, proc)
+            if status == "pipe_closed":
+                return
+    except:  # This task will raise an exception when the loop stops
+        return
+
+
+def handle_outputs_helper(output, socket_id, output_type, lock, event_loop, container_stdout):
+    """ Simple helper to launch the handle_outputs function in its own thread using its own asyncio loop.
+    Used only when both containers are not on a shared kernel """
+    loop = asyncio.new_event_loop()
```

Comment:
Here you run a new loop with a single task, that do not really rely on async code, in a thread. Wouldn't it be more simple to run sync code in the main loop executor ? I may be mistaken about ``write_stdout`` function yet.

Reply:
write_stdout requires to drain the asyncio stream after writing. This is why a loop was required (to await the drain).
I improved it a little bit (having a single loop for stdout and stderr instead of one for each).
Still, I believe the perfect implementation would be to have 3 coroutines (handle stdin reading at agent message, handle_stdout and handle_stderr reading at process outputs) all running on the main loop without using threads.

While this might be improved in the future by someone else (sadly my asyncio comprehension is too limited to achieve it) I believe the current implementation is still "good enough" :/